crates/rpc-types/src/eth/block.rs

```diff
@@ -738,6 +738,30 @@ impl<'de> Deserialize<'de> for BlockId {
     }
 }
 
+/// Error thrown when parsing a [BlockId] from a string.
+#[derive(Debug, thiserror::Error)]
+#[error("failed to parse {input:?} as a number: {parse_int_error} or hash: {hex_error}")]
+pub struct ParseBlockIdError {
+    input: String,
+    parse_int_error: ParseIntError,
+    hex_error: alloy_primitives::hex::FromHexError,
+}
+
+impl FromStr for BlockId {
+    type Err = ParseBlockIdError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match u64::from_str(s) {
+            Ok(val) => Ok(BlockId::Number(BlockNumberOrTag::Number(val))),
+            Err(parse_int_error) => match B256::from_str(s) {
+                Ok(val) => Ok(BlockId::Hash(val.into())),
+                Err(hex_error) => {
+                    Err(ParseBlockIdError { input: s.to_string(), parse_int_error, hex_error })
+                }
+            },
```

Comment:
am I just missing it or does this not cover the tags? only hash and number

Reply:
yeah this is how ethers handles it—we can cover the tags too.